crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,440 @@
+use std::{
+    collections::HashMap,
+    time::Instant,
+};
+
+use fuel_core_types::{
+    fuel_tx::{
+        input::{
+            coin::{
+                CoinPredicate,
+                CoinSigned,
+            },
+            contract::Contract,
+            message::{
+                MessageCoinPredicate,
+                MessageCoinSigned,
+                MessageDataPredicate,
+                MessageDataSigned,
+            },
+        },
+        ContractId,
+        Input,
+        Output,
+        TxId,
+        UtxoId,
+    },
+    services::txpool::PoolTransaction,
+};
+use petgraph::{
+    graph::NodeIndex,
+    prelude::StableDiGraph,
+    visit::EdgeRef,
+};
+
+use crate::{
+    collision_manager::CollisionReason,
+    error::Error,
+    ports::TxPoolDb,
+};
+
+use super::{
+    RemovedTransactions,
+    Storage,
+    StorageData,
+};
+
+pub struct GraphStorage {
```

Comment:
The number of transaction is limited in the pool and so should limit the caches is there is a bug it could not be the case so yeah it might be interesting to add more safety here. Opening an issue.